lib/utils/aws/s3.go

```diff
@@ -62,3 +67,59 @@ func ConvertS3Error(err error, args ...interface{}) error {
 	}
 	return err
 }
+
+// s3V2FileWriter can be used to upload data to s3 via io.WriteCloser interface.
+type s3V2FileWriter struct {
+	// uploadFinisherErrChan is used to wait for completed upload as well as
+	// sending error message.
+	uploadFinisherErrChan <-chan error
+	pipeWriter            *io.PipeWriter
+	pipeReader            *io.PipeReader
+}
+
+func NewS3V2FileWriter(ctx context.Context, s3Client *s3v2.Client, bucket, key string, uploaderOptions []func(*managerV2.Uploader), putObjectInputOptions ...func(*s3v2.PutObjectInput)) (*s3V2FileWriter, error) {
```

Comment:
Can we actually add a test to ensure we will always wait for uploads to finish?

Reply:
added in https://github.com/gravitational/teleport/pull/28442/commits/73aaba37e4b383a8279d60b6a991798a84e0a13a